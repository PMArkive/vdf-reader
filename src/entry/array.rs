use super::Entry;
use serde::{Deserialize, Serialize};
use std::ops::{Deref, DerefMut};

/// An array of entries (items that have the same key).
#[derive(Clone, PartialEq, Eq, Debug, Deserialize, Serialize)]
#[serde(transparent)]
pub struct Array(Vec<Entry>);

impl From<Vec<Entry>> for Array {
    fn from(value: Vec<Entry>) -> Self {
        Array(value)
    }
}
impl From<Entry> for Array {
    fn from(value: Entry) -> Self {
        Array(vec![value])
    }
}

impl From<Array> for Entry {
    fn from(array: Array) -> Self {
        Entry::Array(array)
    }
}

impl Deref for Array {
    type Target = Vec<Entry>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for Array {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}
