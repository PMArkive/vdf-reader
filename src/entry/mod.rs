mod array;
mod statement;
mod table;
mod value;

use crate::error::{ParseEntryError, ParseItemError, ParseStringError};
use crate::Item;
pub use array::Array;
pub use statement::Statement;
use std::any::type_name;
use std::collections::HashMap;
use std::fmt::Formatter;
use std::slice;
pub use table::Table;
pub use value::Value;

/// The kinds of entry.
#[derive(Clone, PartialEq, Eq, Debug, Serialize)]
#[serde(untagged)]
pub enum Entry {
    /// A table.
    Table(Table),

    /// An array (entries with the same key).
    Array(Array),

    /// A value.
    Value(Value),

    /// A statement (the values starting with #).
    Statement(Statement),
}

impl From<Item<'_>> for Entry {
    fn from(item: Item) -> Self {
        match item {
            Item::Item { content, .. } => Entry::Value(content.into()),
            Item::Statement { content, .. } => Entry::Statement(content.into()),
        }
    }
}

impl Entry {
    /// Lookup an entry with a path.
    pub fn lookup<S: AsRef<str>>(&self, path: S) -> Option<&Entry> {
        let mut current = self;

        for name in path.as_ref().split('.') {
            if let Some(entry) = current.get(name.trim()) {
                current = entry;
            } else {
                return None;
            }
        }

        Some(current)
    }

    /// Try to get the named entry.
    pub fn get<S: AsRef<str>>(&self, name: S) -> Option<&Entry> {
        match self {
            Entry::Table(value) => value.get(name.as_ref()),

            Entry::Array(value) => name
                .as_ref()
                .parse::<usize>()
                .ok()
                .and_then(|i| value.get(i)),

            _ => None,
        }
    }

    /// Try to convert the entry to the given type.
    pub fn to<T: ParseItem>(self) -> Result<T, ParseEntryError> {
        T::from_entry(self)
    }

    /// Try to take the entry as a table.
    pub fn as_table(&self) -> Option<&Table> {
        if let Entry::Table(value) = self {
            Some(value)
        } else {
            None
        }
    }

    /// Try to take the entry as a slice.
    pub fn as_slice(&self) -> Option<&[Entry]> {
        if let Entry::Array(value) = self {
            Some(value.as_slice())
        } else {
            unsafe { Some(slice::from_raw_parts(self, 1)) }
        }
    }

    /// Try to take the entry as a statement.
    pub fn as_statement(&self) -> Option<&Statement> {
        if let Entry::Statement(value) = self {
            Some(value)
        } else {
            None
        }
    }

    /// Try to take the entry as a value.
    pub fn as_value(&self) -> Option<&Value> {
        if let Entry::Value(value) = self {
            Some(value)
        } else {
            None
        }
    }

    /// Try to take the entry as a string.
    pub fn as_str(&self) -> Option<&str> {
        match self {
            Entry::Value(value) => Some(value),

            Entry::Statement(value) => Some(value),

            _ => None,
        }
    }
}

/// Parsable types.
pub trait ParseItem: Sized {
    /// Try to cast the entry into a concrete type
    fn from_entry(entry: Entry) -> Result<Self, ParseEntryError> {
        let string = match entry.as_str() {
            Some(string) => string,
            None => {
                return Err(ParseEntryError::new(type_name::<Self>(), entry));
            }
        };
        Self::from_str(string).map_err(|e| ParseEntryError::new(e.ty, entry))
    }

    /// Try to cast the item into a concrete type
    fn from_item(item: Item) -> Result<Self, ParseItemError> {
        Self::from_str(item.as_str()).map_err(|e| ParseItemError::new(e.ty, item))
    }

    /// Try to cast the string into a concrete type
    fn from_str(item: &str) -> Result<Self, ParseStringError>;
}

macro_rules! from_str {
	(for) => ();

	(for $ty:ident $($rest:tt)*) => (
		from_str!($ty);
		from_str!(for $($rest)*);
	);

	($ty:ident) => (
		impl ParseItem for $ty {
			fn from_entry(entry: Entry) -> Result<Self, ParseEntryError> {
                let string = match entry.as_str() {
                    Some(string) => string,
                    None => {
                        return Err(ParseEntryError::new(type_name::<Self>(), entry));
                    }
                };
				string.parse::<$ty>().map_err(|_| ParseEntryError::new(type_name::<Self>(), entry))
			}

            fn from_item(item: Item) -> Result<Self, ParseItemError> {
                item.as_str().parse::<$ty>().map_err(|_| ParseItemError::new(type_name::<Self>(), item))
            }

            fn from_str(item: &str) -> Result<Self, ParseStringError> {
                item.parse::<$ty>().map_err(|_| ParseStringError::new(type_name::<Self>(), item))
            }
		}
	);
}

use serde::de::{Error, MapAccess, Visitor};
use serde::{Deserialize, Deserializer, Serialize};
use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6};
from_str!(for IpAddr Ipv4Addr Ipv6Addr SocketAddr SocketAddrV4 SocketAddrV6);
from_str!(for i8 i16 i32 i64 isize u8 u16 u32 u64 usize f32 f64);

impl ParseItem for bool {
    fn from_str(item: &str) -> Result<Self, ParseStringError> {
        match item {
            "0" => Ok(false),
            "1" => Ok(true),
            v => v
                .parse::<bool>()
                .map_err(|_| ParseStringError::new(type_name::<Self>(), item)),
        }
    }
}

impl ParseItem for String {
    fn from_entry(entry: Entry) -> Result<Self, ParseEntryError> {
        match entry {
            Entry::Table(entry) => Err(ParseEntryError::new(
                type_name::<Self>(),
                Entry::Table(entry),
            )),
            Entry::Array(entry) => Err(ParseEntryError::new(
                type_name::<Self>(),
                Entry::Array(entry),
            )),
            Entry::Statement(statement) => Ok(statement.into()),
            Entry::Value(value) => Ok(value.into()),
        }
    }

    fn from_item(item: Item) -> Result<Self, ParseItemError> {
        Ok(item.into_content().into())
    }

    fn from_str(item: &str) -> Result<Self, ParseStringError> {
        Ok(item.into())
    }
}

impl<T: ParseItem> ParseItem for Option<T> {
    fn from_entry(entry: Entry) -> Result<Self, ParseEntryError> {
        T::from_entry(entry).map(Some)
    }

    fn from_item(item: Item) -> Result<Self, ParseItemError> {
        T::from_item(item).map(Some)
    }

    fn from_str(item: &str) -> Result<Self, ParseStringError> {
        T::from_str(item).map(Some)
    }
}

impl<'de> Deserialize<'de> for Entry {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct EntryVisitor;

        impl<'v> Visitor<'v> for EntryVisitor {
            type Value = Entry;

            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
                write!(formatter, "any string like value or group")
            }

            fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(Entry::Value(v.to_string().into()))
            }

            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(Entry::Value(v.to_string().into()))
            }

            fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(Entry::Value(v.to_string().into()))
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: Error,
            {
                if v.starts_with('#') {
                    Ok(Entry::Statement(v.to_string().into()))
                } else {
                    Ok(Entry::Value(v.to_string().into()))
                }
            }

            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
            where
                E: Error,
            {
                if v.starts_with('#') {
                    Ok(Entry::Statement(v.into()))
                } else {
                    Ok(Entry::Value(v.into()))
                }
            }

            fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
            where
                E: Error,
            {
                let v = if v { "1" } else { "0" };
                Ok(Entry::Value(v.to_string().into()))
            }

            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'v>,
            {
                let mut res = HashMap::new();

                while let Some(entry) = map.next_entry()? {
                    res.insert(entry.0, entry.1);
                }

                Ok(Entry::Table(res.into()))
            }
        }

        deserializer.deserialize_any(EntryVisitor)
    }
}

#[cfg(test)]
#[track_caller]
fn unwrap_err<T>(r: Result<T, crate::VdfError>) -> T {
    r.map_err(miette::Error::from).unwrap()
}

#[test]
fn test_serde_entry() {
    use maplit::hashmap;

    let j = r#"1"#;
    assert_eq!(Entry::Value("1".into()), unwrap_err(crate::from_str(j)));

    let j = r#""foo bar""#;
    assert_eq!(
        Entry::Value("foo bar".into()),
        unwrap_err(crate::from_str(j))
    );

    let j = r#"{foo bar}"#;

    assert_eq!(
        Entry::Table(hashmap! {"foo".into() => Entry::Value("bar".into())}.into()),
        unwrap_err(crate::from_str(j))
    );
}
