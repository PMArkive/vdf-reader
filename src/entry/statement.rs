use super::Entry;
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::ops::Deref;

/// A statement.
#[derive(Clone, PartialEq, Eq, Debug, Deserialize, Serialize)]
#[serde(transparent)]
pub struct Statement(String);

impl From<Cow<'_, str>> for Statement {
    fn from(value: Cow<'_, str>) -> Self {
        Statement(value.into())
    }
}

impl From<Statement> for Entry {
    fn from(statement: Statement) -> Self {
        Entry::Statement(statement)
    }
}

impl From<Statement> for String {
    fn from(value: Statement) -> Self {
        value.0
    }
}

impl Deref for Statement {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
