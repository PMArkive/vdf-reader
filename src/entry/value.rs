use super::Entry;
use serde::de::{Error, Visitor};
use serde::{Deserialize, Deserializer, Serialize};
use std::borrow::Cow;
use std::fmt::Formatter;
use std::ops::Deref;

#[derive(Clone, PartialEq, Eq, Debug, Serialize)]
#[serde(transparent)]
pub struct Value(String);

impl From<Cow<'_, str>> for Value {
    fn from(value: Cow<'_, str>) -> Value {
        Value(value.into())
    }
}
impl From<&'_ str> for Value {
    fn from(value: &str) -> Value {
        Value(value.into())
    }
}

impl From<String> for Value {
    fn from(value: String) -> Value {
        Value(value)
    }
}

impl From<Value> for Entry {
    fn from(value: Value) -> Self {
        Entry::Value(value)
    }
}

impl From<Value> for String {
    fn from(value: Value) -> Self {
        value.0
    }
}

impl Deref for Value {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'de> Deserialize<'de> for Value {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct ValueVisitor;

        impl Visitor<'_> for ValueVisitor {
            type Value = String;

            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
                write!(formatter, "any string like value")
            }

            fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(v.to_string())
            }

            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(v.to_string())
            }

            fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(v.to_string())
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(v.into())
            }

            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(v)
            }

            fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
            where
                E: Error,
            {
                Ok(if v { "1".into() } else { "0".into() })
            }
        }

        deserializer.deserialize_str(ValueVisitor).map(Value)
    }
}

#[cfg(test)]
#[track_caller]
fn unwrap_err<T>(r: Result<T, crate::VdfError>) -> T {
    r.map_err(miette::Error::from).unwrap()
}

#[test]
fn test_serde_value() {
    let j = r#"1"#;
    assert_eq!(Value("1".into()), unwrap_err(crate::from_str(j)));

    let j = r#""foo bar""#;
    assert_eq!(Value("foo bar".into()), unwrap_err(crate::from_str(j)));
}
