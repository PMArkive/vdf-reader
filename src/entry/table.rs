use super::{Array, Entry};
use crate::entry::{string_is_array, Statement, Value};
use crate::error::UnknownError;
use crate::event::{EntryEvent, GroupStartEvent};
use crate::{Event, Item, Reader, Result, VdfError};
use serde::de::{DeserializeSeed, MapAccess};
use serde::{Deserialize, Serialize, Serializer};
use std::collections::hash_map;
use std::collections::HashMap;
use std::ops::{Deref, DerefMut};

/// A table of entries.
#[derive(Clone, PartialEq, Eq, Debug, Deserialize, Serialize)]
#[serde(transparent)]
pub struct Table(#[serde(serialize_with = "ordered_map")] HashMap<String, Entry>);

impl From<HashMap<String, Entry>> for Table {
    fn from(value: HashMap<String, Entry>) -> Self {
        Table(value)
    }
}

fn ordered_map<S, K: Ord + Serialize, V: Serialize>(
    value: &HashMap<K, V>,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    use std::collections::BTreeMap;
    let ordered: BTreeMap<_, _> = value.iter().collect();
    ordered.serialize(serializer)
}

fn insert<K: Into<String>, V: Into<Entry>>(map: &mut HashMap<String, Entry>, key: K, value: V) {
    let key = key.into();
    let value = value.into();
    let entry = map.entry(key);
    match entry {
        hash_map::Entry::Vacant(entry) => {
            entry.insert(value);
        }
        hash_map::Entry::Occupied(mut entry) => match entry.get_mut() {
            Entry::Array(ref mut array) => {
                array.push(value);
            }
            _ => {
                let (key, old_value) = entry.remove_entry();
                let mut array = Array::from(old_value);
                array.push(value);
                map.insert(key, Entry::Array(array));
            }
        },
    }
}

impl Table {
    pub fn load_from_str(input: &str) -> Result<Table> {
        let mut reader = Reader::from(input);
        Self::load(&mut reader)
    }

    /// Load a table from the given `Reader`.
    pub fn load(reader: &mut Reader) -> Result<Table> {
        let mut map = HashMap::new();

        while let Some(event) = reader.event() {
            match event? {
                Event::Entry(EntryEvent {
                    key: Item::Item { content: key, .. },
                    value,
                    ..
                }) => {
                    if string_is_array(value.as_str()) {
                        let str = value.as_str();
                        insert(
                            &mut map,
                            key,
                            Array::from_space_separated(&str[1..str.len() - 1]),
                        )
                    } else {
                        insert(&mut map, key, Value::from(value.into_content()))
                    }
                }

                Event::Entry(EntryEvent {
                    key: Item::Statement { content: key, .. },
                    value,
                    ..
                }) => insert(&mut map, key, Statement::from(value.into_content())),

                Event::GroupStart(GroupStartEvent { name, .. }) => {
                    insert(&mut map, name, Table::load(reader)?)
                }

                Event::GroupEnd(_) => break,
            }
        }

        Ok(Table(map))
    }
}

impl From<Table> for Entry {
    fn from(table: Table) -> Self {
        Entry::Table(table)
    }
}

impl From<Table> for HashMap<String, Entry> {
    fn from(table: Table) -> Self {
        table.0
    }
}

impl Deref for Table {
    type Target = HashMap<String, Entry>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for Table {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

pub(crate) struct TableSeq {
    iter: hash_map::IntoIter<String, Entry>,
    next_item: Option<Entry>,
}

impl TableSeq {
    pub(crate) fn new(table: Table) -> Self {
        TableSeq {
            iter: table.0.into_iter(),
            next_item: None,
        }
    }
}

impl<'de> MapAccess<'de> for TableSeq {
    type Error = VdfError;

    fn next_key_seed<K>(&mut self, seed: K) -> std::result::Result<Option<K::Value>, Self::Error>
    where
        K: DeserializeSeed<'de>,
    {
        let (key, value) = match self.iter.next() {
            Some(pair) => pair,
            None => {
                return Ok(None);
            }
        };
        self.next_item = Some(value);
        seed.deserialize(Value::from(key)).map(Some)
    }

    fn next_value_seed<V>(&mut self, seed: V) -> std::result::Result<V::Value, Self::Error>
    where
        V: DeserializeSeed<'de>,
    {
        let item = match self.next_item.take() {
            Some(item) => item,
            None => return Err(UnknownError::from("double take value").into()),
        };

        seed.deserialize(item)
    }
}

#[cfg(test)]
#[track_caller]
fn unwrap_err<T>(r: Result<T, crate::VdfError>) -> T {
    r.map_err(miette::Error::from).unwrap()
}

#[test]
fn test_serde_table() {
    use maplit::hashmap;

    let j = r#"{foo bar}"#;

    assert_eq!(
        Table(hashmap! {"foo".into() => Entry::Value("bar".into())}),
        unwrap_err(crate::from_str(j))
    );
}
