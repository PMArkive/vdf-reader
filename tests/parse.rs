use std::fs::read_to_string;
use test_case::test_case;
use vdf_reader::entry::Table;
use vdf_reader::Reader;

#[test_case("tests/data/concrete.vmt")]
fn test_parse(path: &str) {
    let raw = read_to_string(path).unwrap();
    let mut reader = Reader::from(raw.as_str());
    let parsed = Table::load(&mut reader).expect("failed to parse test data");
    insta::assert_ron_snapshot!(parsed);
}
