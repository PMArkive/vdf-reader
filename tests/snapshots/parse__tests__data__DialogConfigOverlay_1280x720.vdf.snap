---
source: tests/parse.rs
expression: parsed
---
{
  "UserConfigData": {
    "FriendsMainDialog": {
      "FriendPanelSelf": {},
      "FriendsDialogSheet": {
        "FriendsClansPage": {
          "BuddyList": {
            "0_collapsed": "0",
            "1_collapsed": "0",
            "2_collapsed": "0",
            "3_collapsed": "0",
          },
        },
        "FriendsFriendsPage": {
          "BuddyList": {
            "0_collapsed": "0",
            "1_collapsed": "0",
            "2_collapsed": "0",
            "3_collapsed": "0",
            "4_collapsed": "0",
            "5_collapsed": "0",
            "allfriends_collapsed": "0",
          },
        },
      },
      "FriendsState": {},
      "tall": "440",
      "wide": "252",
      "xpos": "1028",
      "ypos": "280",
    },
    "Servers": {
      "DialogServerBrowser.res": {
        "GameTabs": {
          "FavoriteGames": {
            "gamelist": {
              "#ServerBrowser_Bots_hidden": "1",
              "#ServerBrowser_Game_hidden": "0",
              "#ServerBrowser_IPAddress_hidden": "1",
              "#ServerBrowser_Latency_hidden": "0",
              "#ServerBrowser_Map_hidden": "0",
              "#ServerBrowser_Password_hidden": "0",
              "#ServerBrowser_Players_hidden": "0",
              "#ServerBrowser_Secure_hidden": "0",
              "#ServerBrowser_Servers_hidden": "0",
              "sort_column": "#ServerBrowser_Latency",
              "sort_column_asc": "1",
              "sort_column_secondary": "",
              "sort_column_secondary_asc": "1",
            },
          },
          "FriendsGames": {
            "gamelist": {
              "#ServerBrowser_Bots_hidden": "1",
              "#ServerBrowser_Game_hidden": "0",
              "#ServerBrowser_IPAddress_hidden": "1",
              "#ServerBrowser_Latency_hidden": "0",
              "#ServerBrowser_Map_hidden": "0",
              "#ServerBrowser_Password_hidden": "0",
              "#ServerBrowser_Players_hidden": "0",
              "#ServerBrowser_Secure_hidden": "0",
              "#ServerBrowser_Servers_hidden": "0",
              "sort_column": "#ServerBrowser_Latency",
              "sort_column_asc": "1",
              "sort_column_secondary": "",
              "sort_column_secondary_asc": "1",
            },
          },
          "HistoryGames": {
            "gamelist": {
              "#ServerBrowser_Bots_hidden": "1",
              "#ServerBrowser_Game_hidden": "0",
              "#ServerBrowser_IPAddress_hidden": "1",
              "#ServerBrowser_LastPlayed_hidden": "0",
              "#ServerBrowser_Latency_hidden": "0",
              "#ServerBrowser_Map_hidden": "0",
              "#ServerBrowser_Password_hidden": "0",
              "#ServerBrowser_Players_hidden": "0",
              "#ServerBrowser_Secure_hidden": "0",
              "#ServerBrowser_Servers_hidden": "0",
              "sort_column": "#ServerBrowser_LastPlayed",
              "sort_column_asc": "1",
              "sort_column_secondary": "#ServerBrowser_Latency",
              "sort_column_secondary_asc": "1",
            },
          },
          "InternetGames": {
            "gamelist": {
              "#ServerBrowser_Bots_hidden": "1",
              "#ServerBrowser_Game_hidden": "0",
              "#ServerBrowser_IPAddress_hidden": "1",
              "#ServerBrowser_Latency_hidden": "0",
              "#ServerBrowser_Map_hidden": "0",
              "#ServerBrowser_Password_hidden": "0",
              "#ServerBrowser_Players_hidden": "0",
              "#ServerBrowser_Secure_hidden": "0",
              "#ServerBrowser_Servers_hidden": "0",
              "sort_column": "#ServerBrowser_Latency",
              "sort_column_asc": "1",
              "sort_column_secondary": "",
              "sort_column_secondary_asc": "1",
            },
          },
          "LanGames": {
            "gamelist": {
              "#ServerBrowser_Bots_hidden": "1",
              "#ServerBrowser_Game_hidden": "0",
              "#ServerBrowser_IPAddress_hidden": "1",
              "#ServerBrowser_Latency_hidden": "0",
              "#ServerBrowser_Map_hidden": "0",
              "#ServerBrowser_Password_hidden": "0",
              "#ServerBrowser_Players_hidden": "0",
              "#ServerBrowser_Secure_hidden": "0",
              "#ServerBrowser_Servers_hidden": "0",
              "sort_column": "#ServerBrowser_Latency",
              "sort_column_asc": "1",
              "sort_column_secondary": "",
              "sort_column_secondary_asc": "1",
            },
          },
          "SpectateGames": {
            "gamelist": {
              "#ServerBrowser_Bots_hidden": "1",
              "#ServerBrowser_Game_hidden": "0",
              "#ServerBrowser_IPAddress_hidden": "1",
              "#ServerBrowser_Latency_hidden": "0",
              "#ServerBrowser_Map_hidden": "0",
              "#ServerBrowser_Password_hidden": "0",
              "#ServerBrowser_Players_hidden": "0",
              "#ServerBrowser_Secure_hidden": "0",
              "#ServerBrowser_Servers_hidden": "0",
              "sort_column": "#ServerBrowser_Latency",
              "sort_column_asc": "1",
              "sort_column_secondary": "",
              "sort_column_secondary_asc": "1",
            },
          },
        },
        "tall": "720",
        "wide": "1280",
        "xpos": "0",
        "ypos": "0",
      },
    },
    "Steam": {
      "cached": {
        "OverlaySplash.res": {},
      },
    },
  },
}
